File: Algo/Leetcode/061RotateList.cpp
//1. Link the last node to the first. And record length. Move len - k and link to NULL. O(N).
class Solution {
public:
    ListNode* rotateRight(ListNode* head, int k) {
        if(k==0 || !head) return head;
        ListNode* fast = head;
        int count = 1;
        while(fast->next){
            fast = fast->next;
            count++;
        }
        fast->next = head; // make it circle

        k = k % count;
        fast = head;
        for(int i=0; i<count-k-1; i++) fast = fast->next;

        ListNode* new_head = fast->next;
        fast->next = NULL;
        return new_head;
    }
};

File: Algo/Leetcode/061RotateList.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def rotateRight(self, head, k):
        """
        :type head: ListNode
        :type k: int
        :rtype: ListNode
        """
        if head is None:
            return head

        # get list length
        m = head
        l = 1

        while m.next:
            m = m.next
            l += 1
        # circle
        m.next = head
        k = k % l

        m = head
        i = 0
        while i < l - k - 1:  # be careful on this value
            m = m.next
            i += 1

        res = m.next
        m.next = None

        return res
